aptos-move/aptos-gas/src/instr.rs

```diff
@@ -0,0 +1,330 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::gas_meter::{FromOnChainGasSchedule, InitialGasSchedule, ToOnChainGasSchedule};
+use move_binary_format::{
+    errors::{PartialVMError, PartialVMResult},
+    file_format_common::Opcodes,
+};
+use move_core_types::vm_status::StatusCode;
+use std::collections::BTreeMap;
+
+macro_rules! define_gas_parameters_for_instructions {
+    ($([$name: ident, $key: literal $(,)?, $initial: expr $(,)?]),* $(,)?) => {
+        /// Gas parameters for all bytecode instructions.
+        ///
+        /// Note: due to performance considerations, this is represented as a fixed struct instead of
+        /// some other data structures that require complex lookups.
+        #[derive(Debug, Clone)]
+        pub struct InstructionGasParameters {
+            $(pub $name : u64),*
+        }
+
+        impl FromOnChainGasSchedule for InstructionGasParameters {
+            fn from_on_chain_gas_schedule(gas_schedule: &BTreeMap<String, u64>) -> Option<Self> {
+                Some(InstructionGasParameters { $($name: gas_schedule.get(&format!("instr.{}", $key)).cloned()?),* })
+            }
+        }
+
+        impl ToOnChainGasSchedule for InstructionGasParameters {
+            fn to_on_chain_gas_schedule(&self) -> Vec<(String, u64)> {
+                vec![$((format!("instr.{}", $key), self.$name)),*]
+            }
+        }
+
+        impl InstructionGasParameters {
+            pub fn zeros() -> Self {
+                Self {
+                    $($name: 0),*
+                }
+            }
+        }
+
+        impl InitialGasSchedule for InstructionGasParameters {
+            fn initial() -> Self {
+                Self {
+                    $($name: $initial),*
+                }
+            }
+        }
+
+        #[test]
+        fn keys_should_be_unique() {
+            let mut map: BTreeMap<&str, ()> = BTreeMap::new();
+
+            for key in [$($key),*] {
+                assert!(map.insert(key, ()).is_none());
+            }
+        }
+    };
+}
+
+define_gas_parameters_for_instructions!(
+    // nop
+    [nop, "nop", 1],
+    // control flow
+    [ret, "ret", 1],
+    [abort, "abort", 1],
+    [br_true, "br_true", 1],
+    [br_false, "br_false", 1],
+    [branch, "branch", 1],
+    // stack
+    [pop, "pop", 1],
+    [ld_u8, "ld_u8", 1],
+    [ld_u64, "ld_u64", 1],
+    [ld_u128, "ld_u128", 1],
+    [ld_true, "ld_true", 1],
+    [ld_false, "ld_false", 1],
+    [ld_const_base, "ld_const.base", 1],
+    [ld_const_unit, "ld_const.unit", 1],
```

Comment:
I thought about this. I think it's possible but there can always be unexpected things with complex macros so I can't say for sure. I have some rough ideas I can try later.

Maybe another thing we could explore is procedural macro, but I feel like it may not work with upstream types (e.g. `move_stdlib::natives::GasParameters`)